test/e2e_node/device_manager_test.go

```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 		})
 
 	})
+
+	/*
+		This end to end test is to simulate a scenario where after kubelet restart/node
+		reboot application pods requesting devices appear before the device plugin
+		pod exposing those devices as resources.
+
+		The happy path is where after node reboot/ kubelet restart, the device plugin pod
+		appears before the application pod. This PR and this e2e test
+		aims to tackle the scenario where device plugin either does not appear first
+		or doesn't get the chance to re-register itself.
+
+		Since there is no way of controlling the order in which the pods appear after
+		kubelet restart/node reboot, we can't guarantee that the application pod
+		recovers before device plugin pod (the scenario we want to exercise here).
+		If the device plugin pod is recovered before the test pod, we still can
+		meaningfully reproduce the scenario by NOT sending the registration command.
+		To do so sample device plugin is enhanced. For implementation details, refer to:
+		`test/images/sample-device-plugin/sampledeviceplugin.go`. This enhancement
+		allows auto-registration of the plugin to be controlled with the help of an environment
+		variable: AUTO_REGISTER. By default AUTO_REGISTER=true allowing the device plugin
+		to autoregister to kubelet. For this e2e test, we use sample device plugin
+		spec with AUTO_REGISTER=false to first manually register the plugin to allow
+		an application pod (requesting devices) to successfully run on the node
+		followed by kubelet restart where device plugin doesn't register and the
+		application pod fails with admission error.
```

Comment:
Can we achieve the same result by not starting device plugin after the kubelet restart?

Reply:
The device plugin and the application pod are recovered automatically after kubelet restart. We cannot influence the device plugin pod start-up as the order in which the pods are recovered is not in our control.

Reply:
Now it makes sense to me. Thanks for the explanations.

Reply:
No problem!